crates/services/p2p/src/limited_behaviour.rs

```diff
@@ -0,0 +1,275 @@
+use crate::connection_limits::Connections;
+use libp2p::{
+    core::{
+        transport::PortUse,
+        Endpoint,
+        Multiaddr,
+    },
+    identity::PeerId,
+    swarm::{
+        behaviour::ConnectionEstablished,
+        derive_prelude::Either,
+        dummy,
+        AddressChange,
+        ConnectionClosed,
+        ConnectionDenied,
+        ConnectionId,
+        DialFailure,
+        FromSwarm,
+        ListenFailure,
+        NetworkBehaviour,
+        THandler,
+        THandlerInEvent,
+        THandlerOutEvent,
+        ToSwarm,
+    },
+};
+use std::{
+    collections::HashSet,
+    ops::{
+        Deref,
+        DerefMut,
+    },
+    task::{
+        Context,
+        Poll,
+    },
+};
+
+pub struct LimitedBehaviour<Behaviour> {
+    connection_limit: usize,
+    behaviour: Behaviour,
+    connections: Connections,
+    connected_peers: HashSet<PeerId>,
+    known_connections: HashSet<ConnectionId>,
+}
+
+impl<Behaviour> Deref for LimitedBehaviour<Behaviour> {
+    type Target = Behaviour;
+
+    fn deref(&self) -> &Self::Target {
+        &self.behaviour
+    }
+}
+
+impl<Behaviour> DerefMut for LimitedBehaviour<Behaviour> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.behaviour
+    }
+}
+
+impl<Behaviour> LimitedBehaviour<Behaviour> {
+    pub fn new(
+        connection_limit: usize,
+        connections: Connections,
+        behaviour: Behaviour,
+    ) -> Self {
+        Self {
+            connection_limit,
+            behaviour,
+            connections,
+            connected_peers: Default::default(),
+            known_connections: Default::default(),
+        }
+    }
+}
+
+impl<Behaviour> NetworkBehaviour for LimitedBehaviour<Behaviour>
+where
+    Behaviour: NetworkBehaviour,
+{
+    type ConnectionHandler =
+        Either<dummy::ConnectionHandler, Behaviour::ConnectionHandler>;
+    type ToSwarm = Behaviour::ToSwarm;
+
+    fn handle_pending_inbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> Result<(), ConnectionDenied> {
+        Ok(())
+    }
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        let connections = self.connections.lock();
+        let total_connections = connections.established_per_peer.len();
+
+        if connections.reserved_peers.contains(&peer)
+            || self.connected_peers.contains(&peer)
```

Comment:
is there an invariant where `self.connected_peers` will contain the peer when `handle_established_inbound_connection` is called? I'm not sure why a peer will have multiple connections, that seems like an anti-pattern if we have a multiplexer

if it's being cleaned up when they disconnect, in https://github.com/FuelLabs/fuel-core/pull/2743/files#diff-8ea0bda515a7e6ad6977be6e69a2be8df72c00c0aebe4f7c3af939f5ed46109dR185 then we may not need this conditional

Reply:
Yes, even in our tests I had several use cases when we had several connection ids per one peer.

The `total_connections` represents the number of unique peers that are connected to us. And if we started to provide service for `peer_id`, we want to continue to provide service in the case if he opens a new connections.

The total number of connections is limited by `connection_limit::Behaviour` to 3, so you can't exploit it.